File: databutler_widgets/databutler_widgets/datana/widget.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Rohan Bavishi.
# Distributed under the terms of the Modified BSD License.

"""
TODO: Add module docstring
"""
from typing import Callable

from ipywidgets import DOMWidget
from traitlets import Unicode, Bool, List, Dict
from .._frontend import module_name, module_version

IMAGE_LINK = "https://chartio.com/assets/dfd59f/tutorials/charts/grouped-bar-charts/c1fde6017511bbef7ba9bb245a113c07f8ff32173a7c0d742a4e1eac1930a3c5/grouped-bar-example-1.png"


class DatanaExampleWidget(DOMWidget):
    _model_name = Unicode('DatanaExampleModel').tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)
    _view_name = Unicode('DatanaExampleView').tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)

    # Your widget state goes here. Make sure to update the corresponding
    # JavaScript widget state (defaultModelProperties) in widget.ts
    search_box_value = Unicode('').tag(sync=True)
    search_options = List([]).tag(sync=True)
    search_selected = Bool(False).tag(sync=True)

    graphs_generated = List([]).tag(sync=True)

    highlighted_graph = Dict({}).tag(sync=True)
    unchecked_mods_list = List([]).tag(sync=True)

    def __init__(self):
        super().__init__()

    def callback_method(self, callback_fn: Callable, callback_var: str):
        # the callback function never takes any arguments
        def wrapped_callback(*args, **kwargs):
            callback_fn()
        self.observe(wrapped_callback, callback_var)

File: databutler_widgets/src/widget.ts
// Copyright (c) Rohan Bavishi
// Distributed under the terms of the Modified BSD License.

import {
  DOMWidgetModel,
  DOMWidgetView,
  ISerializers,
} from '@jupyter-widgets/base';
import ReactWidget from "./ReactWidget"
import React from 'react';
import ReactDOM from 'react-dom';

import { MODULE_NAME, MODULE_VERSION } from './version';

// Import the CSS
import '../css/widget.css';

// Your widget state goes here. Make sure to update the corresponding
// Python state in widget.py
const defaultModelProperties = {
  search_box_value: '',
  search_selected: false,
  search_options: [],
  graphs_generated: [],
  highlighted_graph: {},
  mods_list: []
}

export type WidgetModelState = typeof defaultModelProperties

export class DatanaExampleModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: DatanaExampleModel.model_name,
      _model_module: DatanaExampleModel.model_module,
      _model_module_version: DatanaExampleModel.model_module_version,
      _view_name: DatanaExampleModel.view_name,
      _view_module: DatanaExampleModel.view_module,
      _view_module_version: DatanaExampleModel.view_module_version,
      ...defaultModelProperties
    };
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
    // Add any extra serializers here
  };

  static model_name = 'DatanaExampleModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'DatanaExampleView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class DatanaExampleView extends DOMWidgetView {
  render() {
    this.el.classList.add('custom-widget');

    const component = React.createElement(ReactWidget, {
      model: this.model,
    });
    ReactDOM.render(component, this.el);
  }
}

export class VizsmithExampleModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: VizsmithExampleModel.model_name,
      _model_module: VizsmithExampleModel.model_module,
      _model_module_version: VizsmithExampleModel.model_module_version,
      _view_name: VizsmithExampleModel.view_name,
      _view_module: VizsmithExampleModel.view_module,
      _view_module_version: VizsmithExampleModel.view_module_version,
      ...defaultModelProperties
    };
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
    // Add any extra serializers here
  };

  static model_name = 'VizsmithExampleModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'VizsmithExampleView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class VizsmithExampleView extends DOMWidgetView {
  render() {
    this.el.classList.add('custom-widget');

    const component = React.createElement(ReactWidget, {
      model: this.model,
    });
    ReactDOM.render(component, this.el);
  }
}
